### PR TITLE
Add map-based storage as well as array-based storage

### PR DESCRIPTION
**Pull request by hhexo**
Hi markomanninen,

currently the storage functionality in the plugin creates an array. This is very useful in many cases, however if you then want to use only one of the elements you can only search by index. It would be nice if each stored value could optionally be indexed by a "key" which could be a portion of the regexp (one of the groups, typically).

For example, if I had a regexp pattern like this:

```
<!-- my_regex code="(.*?)" value="(.*?)" -->
```

And I wrote something like this in my markdown:

```
<!-- my_regex code="my_code" value="blah blah blah" -->
```

Then I might want to store the matched text in a map keyed by the "code", so I can retrieve it with something like:

```
{{ book.my_variable["my_code"] }}
```

And this would print "blah blah blah".

My pull request adds this feature.

Basically, you can now provide a `key` property under `store`. If you do, the variable where the matches are stored is treated as a map rather than an array. The `key` string is processed (supporting `_INDEX_`, `_PAGE_LEVEL_`, the regexp groups, and so on) and the result is used as the key in the map. The value is instead the usual `store.substitute`.

This makes it possible to have a sort of "database" in one of your markdown files, and extract individual rows from the "database" in other markdown files by just indexing a variable. It's pretty neat.

I've made this change locally to your plugin for one of my gitbooks, and I'm very happy it works as intended.

I hope you like it too; if so, feel free to accept the pull request.

### CODE REVIEW

**Comment by markomanninen** [comment]
Fantastic! I had this on my mind when doing key based citations. I need to test it. Couple of things comes to my mind. What is the advantage of doing database this way rather than putting variables to book.json? We should also modify readme file and give an example how this feature can be used, like other examples. Is there a possibility to write over other variables?

Because variables can be stored on book.json in centralized way, I think one of the benefits of initializing them on the book pages is, that variables are then relative to the book structure. Or what do you think @hhexo ?

**Comment by hhexo** [comment]
You make a good point about the readme file, I should provide an example.
I'm a bit busy this weekend but I'll try to get to it as soon as possible.

The main reason why I prefer to do the "database" this way rather than with the book.json variables is because I may have a _lot_ of data and book.json would become quite large. I'd rather write my data in a markdown file with comments and regexps to match those. Or even, I could use regexps to match comma-separated values in a CSV file.

My use case is... writing a role-playing game rulebook. :) 
My RPG system has a long list of "traits" (like, say, a +1 bonus to some ability) which can be obtained by the character either individually or in sort of "packages" which include several "traits". There is a one-to-many relation between traits and packages because a trait can be contained in more than one package.

The character's race/species is a "package": if you're an Elf, you're good at seeing in the dark (bonus to perception) and you're good with a bow (bonus to archery). But the character's profession is also a package, and you could be a Hunter who is also good with the bow (same bonus to archery).

So what I want to do is to have the list of traits in one big markdown file which also acts as the traits index, and then I want to reference individual traits from the parts of the rulebook where I'm describing races, professions, and so on. The full data for the trait is in one place which acts as the source of truth, and it gets copied in the documents when they are generated. It's basically the DRY principle applied to writing books instead of software. :)

In response to another of your questions, my change does affect variables that are already in book.json. But so does the existing behaviour: if you try to store to a variable name that exists, and that variable isn't an array, you get an error because you can't push() into it.

**Comment by hhexo** [comment]
Hi markomanninen,

I have updated the pull request with an example at the end of the README.md. I have reused the citations idea, and presented a different way of doing it using the map-based functionality.

**Comment by markomanninen** [comment]
Perfectly shows pretty complex yet powerful ways to configure the plugin and use it for different purposes. Role game rulebook storage, key based citations,... Thank you for your contribution and detailed explanations, I should be able to merge it this weekend.
